### PR TITLE
#3: add random civ choice

### PR DESCRIPTION
**Pull request by T90Watcher**

### CODE REVIEW

File: src/models/Validator.ts [hunk 1]
```diff
@@ -19,6 +19,9 @@ export class Validator {
         if (Util.isPlayerEvent(message)) {
             const playerEvent: PlayerEvent = message as PlayerEvent;
 
+            if (Util.isRandomCiv(message.civilisation)) {
```
Comment (HSZemi) [inline src/models/Validator.ts]: I feel like the random civ replacement should not happen inside the Validator, but before

Comment (T90Watcher) [reply src/models/Validator.ts]: I moved it to server.ts

File: src/models/Util.ts [hunk 2]
```diff
@@ -78,6 +80,28 @@ export const Util = {
...
-    }
+    },
+
+    isRandomCiv(civilisation: Civilisation): boolean {
+        return civilisation.name.toUpperCase() === "RANDOM";
+    },
 
+    setRandomCivilisation(playerEvent: PlayerEvent, draftId: string, draftStore: DraftsStore) {
+        const maxCivilisationIndex = Civilisation.ALL.length - 1;
+        const randomCivIndex = Math.floor(Math.random() * maxCivilisationIndex);
```
Comment (HSZemi) [inline src/models/Util.ts]: Math.random() yields [0,1),  so `Math.floor(Math.random() * array.length)` yields a valid array index – the `- 1` leads to the last civ never being picked

File: src/models/Util.ts [hunk 3]
```diff
@@ -78,6 +80,28 @@ export const Util = {
...
+        playerEvent.civilisation = Civilisation.ALL[randomCivIndex];
+
+        const errors = Validator.checkAllValidations(draftId, draftStore, playerEvent);
+
+        if (errors.length === 0) {
+            // random civ is set correctly.
+            playerEvent.civilisation.isRandomlyChosenCiv = true;
+            return;
+        } else {
+            console.log("Errors:", errors, "Randomly chosen civ:", Civilisation.ALL[randomCivIndex]);
```
Comment (HSZemi) [inline src/models/Util.ts]: Remove this debugging output, it would print too much in production

File: src/models/Util.ts [hunk 4]
```diff
@@ -78,6 +80,28 @@ export const Util = {
...
+            // recursively try to set random civ
+            playerEvent.civilisation = Civilisation.RANDOM;
```
Comment (HSZemi) [inline src/models/Util.ts]: It is a bit unfortunate that we have to reset the event's civilisation to `RANDOM` again here

File: src/server.ts [hunk 5]
```diff
@@ -103,6 +104,9 @@ io.on("connection", (socket: socketio.Socket) => {
 
     socket.on("act", (message: PlayerEvent, fn: (retval: any) => void) => {
         console.log(message);
+        if (Util.isRandomCiv(message.civilisation)) {
+            Util.setRandomCivilisation(message, draftId, draftsStore);
+        }
```
Comment (HSZemi) [inline src/server.ts]: I would replace this block with a single method call that
- checks if the civilisation is `RANDOM`
- replaces it with a valid option if that is the case
- returns the new `PlayerEvent`

Comment (HSZemi) [reply src/server.ts]: could look like
```javascript
message = Util.setRandomCivilisationIfNeeded(message, draftid, draftsStore);
```

File: src/models/Civilisation.ts [hunk 6]
```diff
@@ -58,6 +59,7 @@ class Civilisation {
     public static readonly TEUTONS:Civilisation = new Civilisation(Name.TEUTONS, GameVersion.AOK);
     public static readonly TURKS:Civilisation = new Civilisation(Name.TURKS, GameVersion.AOK);
     public static readonly VIKINGS:Civilisation = new Civilisation(Name.VIKINGS, GameVersion.AOK);
+    public static readonly RANDOM: Civilisation = new Civilisation(Name.RANDOM, GameVersion.AOK);
```
[the remark this answers is not in the document]
Comment (T90Watcher) [reply src/models/Civilisation.ts]: Civs with GameVersion.TECHNICAL have hidden names on the UI

[the remark this answers is not in the document]
Comment (HSZemi) [reply src/models/Civilisation.ts]: Good point.

File: src/models/Util.ts [hunk 7]
```diff
@@ -78,6 +80,33 @@ export const Util = {
...
         return civilisation.gameVersion === GameVersion.TECHNICAL;
-    }
+    },
 
+    isRandomCivilisation(civilisation: Civilisation): boolean {
+        return civilisation.name.toUpperCase() === "RANDOM";
+    },
+
+    getRandomCivilisation(): Civilisation {
+        const maxCivilisationIndex = Civilisation.ALL.length;
```
Comment (HSZemi) [inline src/models/Util.ts]: The Preset could have a different set of available civilisations, for example only civs from AoK/AoC. We should only try civs from the allowed set of civilisations in `draft.preset.civilisations`

Comment (HSZemi) [reply src/models/Util.ts]: We could start off with a copy of `draft.preset.civilisations` in `setRandomCivilisationIfNeeded()`, and then remove any civ that we tried. This way we do not have to check a civilisation twice that we already rejected before.

File: src/models/Util.ts [hunk 8]
```diff
@@ -78,6 +80,35 @@ export const Util = {
...
+    },
 
+    isRandomCivilisation(civilisation: Civilisation): boolean {
+        return civilisation.name.toUpperCase() === "RANDOM";
+    },
+
+    getRandomCivilisation(civilisationsList: Civilisation[]): Civilisation {
+        const maxCivilisationIndex = civilisationsList.length;
+        const randomCivIndex = Math.floor(Math.random() * maxCivilisationIndex);
+        // remove the random civilisation from the list and return it
```
Comment (HSZemi) [inline src/models/Util.ts]: Remove this redundant comment

File: src/models/Util.ts [hunk 9]
```diff
@@ -78,6 +80,35 @@ export const Util = {
...
+            const randomCiv = Util.getRandomCivilisation(civilisationsList);
+            const playerEventForValidation = new PlayerEvent(playerEvent.player, playerEvent.actionType, randomCiv);
+            const errors = Validator.checkAllValidations(draftId, draftStore, playerEventForValidation);
+            if (errors.length === 0) {
+                // random civ is set correctly.
+                playerEvent.civilisation = randomCiv;
+                playerEvent.civilisation.isRandomlyChosenCiv = true;
+                return playerEvent;
+            } else {
+                // recursively try to set random civ
```
Comment (HSZemi) [inline src/models/Util.ts]: Remove this redundant comment

File: src/models/Util.ts [hunk 10]
```diff
@@ -78,6 +80,35 @@ export const Util = {
...
+    },
+
+    setRandomCivilisationIfNeeded(playerEvent: PlayerEvent, draftId: string,
+                                  draftStore: DraftsStore, civilisationsList: Civilisation[]): PlayerEvent {
+        if (Util.isRandomCivilisation(playerEvent.civilisation)) {
+            const randomCiv = Util.getRandomCivilisation(civilisationsList);
+            const playerEventForValidation = new PlayerEvent(playerEvent.player, playerEvent.actionType, randomCiv);
+            const errors = Validator.checkAllValidations(draftId, draftStore, playerEventForValidation);
+            if (errors.length === 0) {
+                // random civ is set correctly.
```
Comment (HSZemi) [inline src/models/Util.ts]: Is this comment necessary?

**Pull request merged**